sql/src/main/java/io/crate/metadata/sys/TableHealthService.java

```diff
@@ -0,0 +1,250 @@
+            shardsInfo.replicating += shardCount;
+        }
+    }
+
+    private static class HealthResultReceiver implements ResultReceiver {
+
+        private static final Logger LOGGER = Loggers.getLogger(TableHealthService.HealthResultReceiver.class);
+
+        private final CompletableFuture<Map<TablePartitionIdent, ShardsInfo>> result;
+        private final Map<TablePartitionIdent, ShardsInfo> tables = new HashMap<>();
+
+        HealthResultReceiver(CompletableFuture<Map<TablePartitionIdent, ShardsInfo>> result) {
+            this.result = result;
+        }
+
+        @Override
+        public void setNextRow(Row row) {
+            TablePartitionIdent ident = new TablePartitionIdent(
+                BytesRefs.toBytesRef(row.get(0)), BytesRefs.toBytesRef(row.get(1)), BytesRefs.toBytesRef(row.get(2)));
+            ShardsInfo shardsInfo = tables.getOrDefault(ident, new ShardsInfo());
+            String routingState = BytesRefs.toString(row.get(3));
+            boolean primary = (boolean) row.get(4);
+            String relocatingNode = BytesRefs.toString(row.get(5));
+            long cnt = (long) row.get(6);
+
+            collectShardInfo(shardsInfo, routingState, primary, cnt, relocatingNode);
+            tables.put(ident, shardsInfo);
+        }
+
+        @Override
+        public void allFinished(boolean interrupted) {
+            result.complete(tables);
+        }
+
+        @Override
+        public void fail(@Nonnull Throwable t) {
+            LOGGER.error("error retrieving tables health", t);
+            result.completeExceptionally(t);
+        }
+
+        @Override
+        public void batchFinished() {
+        }
+
+        @Override
+        public CompletableFuture<?> completionFuture() {
+            return result;
+        }
+    }
+
+    @VisibleForTesting
+    static class TablePartitionIdent {
+        private final BytesRef tableName;
+        private final BytesRef tableSchema;
+        @Nullable
+        private final BytesRef partitionIdent;
+
+        TablePartitionIdent(BytesRef tableName, BytesRef tableSchema, @Nullable BytesRef partitionIdent) {
+            this.tableName = tableName;
+            this.tableSchema = tableSchema;
+            this.partitionIdent = partitionIdent;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            TablePartitionIdent that = (TablePartitionIdent) o;
+            return Objects.equals(tableName, that.tableName) &&
+                   Objects.equals(tableSchema, that.tableSchema) &&
+                   Objects.equals(partitionIdent, that.partitionIdent);
+        }
+
+        @Override
+        public int hashCode() {
+
+            return Objects.hash(tableName, tableSchema, partitionIdent);
+        }
+    }
+
+    @VisibleForTesting
+    static class ShardsInfo {
+        long activePrimaries = 0;
```

Comment:
could all those be `int` instead of `long`

Reply:
the count aggregate of the statement will return a long, would have to get the intValue...

Reply:
ah yes, actually wanted to ask there (missed the ?), thx!